onnxruntime/test/contrib_ops/ngram_test.cc

```diff
@@ -0,0 +1,565 @@
+}
+
+TEST(ContribOpTest, Ngram_String_TF_BatchOnlyBigrams_Skip5) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , Min=Max=2, weights empty, string
+  InitTestAttr(test, "TF", 2, 2, 5,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {},
+               {"two", "three", "five", "four",                     //1-grams
+                "five", "six", "seven", "eight", "six", "seven"});  //bi-grams
+
+  std::vector<int64_t> dims{2, 6};
+  std::vector<std::string> input{"one", "one", "three", "three", "three", "seven", "eight",
+                                 "six", "seven", "five", "six", "eight"};
+  test.AddInput<std::string>("T", dims, input);
+
+  std::vector<int64_t> out_dims{2, 7};
+  // No 1-grams but Skip is 5 so we manage to count 3
+  // occurrences of [7,8]
+  std::vector<float> output = {0, 0, 0, 0, 0, 0, 0,
+                               0, 0, 0, 0, 1, 1, 1};
+  test.AddOutput<float>("Y", out_dims, output);
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess);
+}
+
+TEST(ContribOpTest, Ngram_Int32_TF_UniAndBigrams_Skip5) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=5, , Min=1, Max=2, weights empty, int32
+  InitTestAttr(test, "TF", 1, 2, 5,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {2, 3, 5, 4,         //1-grams
+                5, 6, 7, 8, 6, 7},  //bi-grams
+               {});
+
+  std::vector<int64_t> dims{12};
+  std::vector<int32_t> input = {1, 1, 3, 3, 3, 7, 8, 6, 7, 5, 6, 8};
+  test.AddInput<int32_t>("T", dims, input);
+
+  std::vector<int64_t> out_dims{7};
+  // We consider both 1-grams and 2-grams so get all the counts here
+  std::vector<float> output = {0, 3, 1, 0, 1, 3, 1};
+  test.AddOutput<float>("Y", out_dims, output);
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess);
+}
+
+TEST(ContribOpTest, Ngram_Int32_TF_BatchUniAndBigrams_Skip5) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=5, , Min=1, Max=2, weights empty, int32
+  InitTestAttr(test, "TF", 1, 2, 5,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {2, 3, 5, 4,         //1-grams
+                5, 6, 7, 8, 6, 7},  //bi-grams
+               {});
+
+  std::vector<int64_t> dims{2, 6};
+  std::vector<int32_t> input = {1, 1, 3, 3, 3, 7,
+                                8, 6, 7, 5, 6, 8};
+  test.AddInput<int32_t>("T", dims, input);
+
+  std::vector<int64_t> out_dims{2, 7};
+  // Counts are now per row (batch)
+  std::vector<float> output = {0, 3, 0, 0, 0, 0, 0,
+                               0, 0, 1, 0, 1, 1, 1};
+  test.AddOutput<float>("Y", out_dims, output);
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess);
+}
+
+TEST(ContribOpTest, Ngram_String_TF_UniAndBigrams_Skip5) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , Min=1, Max=2, weights empty, string
```

Comment:
>1 - 2, s=0, , Min=1, Max=2, weights empty, string [](start = 5, length = 49)

A lot of these comments in the tests are incorrect. Please update or remove.

Reply:
I have updated the comments last time I addressed comments and they are correct. In particular, the highlighted comment is correct. Please, be specific here.

---
In reply to: [244894221](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r244894221) [](ancestors = 244894221)